lib/Sema/TypeCheckType.cpp

```diff
@@ -5902,42 +5902,68 @@ TypeResolver::resolveExistentialType(ExistentialTypeRepr *repr,
   if (constraintType->hasError())
     return ErrorType::get(getASTContext());
 
+  auto contextualConstraintType = constraintType;
+
+  // Map the constraint type into context before performing validity checks.
+  switch (this->resolution.getStage()) {
+  case TypeResolutionStage::Structural:
+    // In the structural stage we are not allowed to compute the generic
+    // signature because we may already be in the middle of computing it.
+    // If validity is questionable, i.e., the constraint type is a type
+    // parameter, assume the constraint type is valid.
+    if (constraintType->isTypeParameter()) {
+      return ExistentialType::get(constraintType);
+    }
+
+    break;
+  case TypeResolutionStage::Interface:
+    contextualConstraintType = this->resolution.getGenericSignature()
+                                   .getGenericEnvironment()
+                                   ->mapTypeIntoContext(constraintType);
+
+    // If there was a problem with the type, don't emit another diagnostic.
+    if (contextualConstraintType->hasError()) {
+      return ErrorType::get(getASTContext());
+    }
+
+    break;
+  }
+
+  if (contextualConstraintType->isConstraintType()) {
+    return ExistentialType::get(constraintType);
```

Comment:
`constraintType` is still going to be a type parameter in this case. Perhaps you want `contextualConstraintType->mapTypeOutOfContext()`?

Reply:
Shouldn’t it be fine (as long as the contextual type *is* a constraint type)?

Reply:
it would be nice if, eg `ExistentialType::get()` could assert that the given type actually is a constraint type.